### PR TITLE
7908 - Functional Tests for the Cosmosdb Worker Function

### PR DESCRIPTION
**Pull request by AndyHale**
### 📲 What

Add functional tests to the CosmosDB Worker Azure Function.  These tests check that when items are created and updated in the Cosmos Container, that messages are sent to the Azure Service Bus Topic to raise events for these changes.

### 🤔 Why

These tests serve as an example for the templates and will be run in the pipeline for quality assurance.

### 🛠 How

An example BDDfy User Story has been created with the following scenarios: -

* When an item, with valid data, is created in CosmosDB, then a message is sent to Azure Service Bus to raise an event.
* When an item, with valid data, is updated in CosmosDB then a message is sent to Azure Service Bus to raise an event.
* When an item, with invalid data, is created in CosmosDB, then no message is sent to Azure Service Bus.

### 👀 Evidence

The screenshot below shows all three tests passing: -

image

### 🕵️ How to test

Ensure all tests are passing and follow each scenario through to ensure steps behave as expected.

### ✅ Acceptance criteria Checklist

[ ] Code peer reviewed?
[ ] Documentation has been updated to reflect the changes?
[ ] Passing all automated tests, including a successful deployment?
[ ] Passing any exploratory testing?
[ ] Rebased/merged with latest changes from development and re-tested?
[ ] Meeting the Coding Standards?

### CODE REVIEW

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.sln [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿
```
Comment (sharonrussell) [inline src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.sln:1]: Is there two sln files?

**Review by sharonrussell — CHANGES_REQUESTED** [review]
Just wondering why there are two .sln files

**Review by sharonrussell — APPROVED** [review]

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.sln (hunk 1 above)
Comment (AndyHale) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.sln:1]: Yes, there were!  Thanks very much.
It looks like one was left behind after moving the solution. 
Cleaned up now.

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Drivers/CosmosDbDriver.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        catch (Exception e)
+        {
+            Console.WriteLine("Error : " + e.Message);
+            return false;
+        }
+    }
+
+
+    /// <summary>
+    /// Replace an existing item in a CosmosDB container
```
Comment (FedericaCola) [inline src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Drivers/CosmosDbDriver.cs:97]: Is this "Delete" ?

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Tests/Stories/ExampleStory.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    public async Task DisposeAsync()
+    {
+        await _exampleSteps.DeleteItemsCreatedInCosmosDb();
+    }
+
+
+    [Fact]
+    public void Confirm_when_create_valid_item_in_CosmosDb_then_event_is_published_to_Service_Bus()
```
[the remark this answers is not in the document]
Comment (FedericaCola) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Tests/Stories/ExampleStory.cs:28]: Seems like the deletion if an Item in Cosmos is not logged at all, that's why there are no tests for it

[the remark this answers is not in the document]
Comment (AndyHale) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Tests/Stories/ExampleStory.cs:28]: The [documentation for the CosmosDBTrigger ](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-trigger) that the change feed won't trigger for deletes.

> The Azure Cosmos DB Trigger uses the [Azure Cosmos DB change feed](https://learn.microsoft.com/en-us/azure/cosmos-db/change-feed) to listen for inserts and updates across partitions. **The change feed publishes new and updated items, not including updates from deletions**.

As this wasn't implemented, we didn't create a test for it.  Perhaps worth extending our example thought to show how we could handle deletions.

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Drivers/CosmosDbDriver.cs (hunk 2 above)
Comment (AndyHale) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Drivers/CosmosDbDriver.cs:97]: Yes it is, thanks very much for spotting the typo.
Will correct.

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Tests/Stories/ExampleStory.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackMidd-Amido) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/Tests/Stories/ExampleStory.cs:28]: No, in that case this is all OK :)

**Review by JackMidd-Amido — APPROVED** [review]

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.csproj [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    <Nullable>enable</Nullable>
+
+    <IsPackable>false</IsPackable>
+    <IsTestProject>true</IsTestProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Azure.Messaging.ServiceBus" Version="7.17.5" />
+    <PackageReference Include="Microsoft.Azure.Cosmos" Version="3.44.0" />
+    <PackageReference Include="coverlet.collector" Version="6.0.0" />
```
Comment (sharonrussell) [inline src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.csproj:15]: needs to be version 3.2.0

**Review by sharonrussell — CHANGES_REQUESTED** [review]
Just one comment about coverlet.collector

File: src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.csproj (hunk 4 above)
Comment (AndyHale) [reply src/func-cosmosdb-worker/src/tests/Functional/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests/xxENSONOxx.xxSTACKSxx.Worker.FunctionalTests.csproj]: Downgraded. 👍

**Review by sharonrussell — APPROVED** [review]

**Pull request merged**